service/controller/resource/clusterdependents/resource.go

```diff
@@ -0,0 +1,115 @@
+package clusterdependents
+
+import (
+	"context"
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/giantswarm/operatorkit/controller/context/finalizerskeptcontext"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"sigs.k8s.io/cluster-api-provider-azure/api/v1alpha3"
+	capiv1alpha3 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	expcapiv1alpha3 "sigs.k8s.io/cluster-api/exp/api/v1alpha3"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/giantswarm/azure-operator/v4/pkg/label"
+	"github.com/giantswarm/azure-operator/v4/service/controller/key"
+)
+
+const (
+	// Name is the identifier of the resource.
+	Name = "clusterdependents"
+)
+
+type Config struct {
+	CtrlClient client.Client
+	Logger     micrologger.Logger
+}
+
+// Resource ensures that there are no orphaned dependent AzureCluster or MachinePool CRs.
+type Resource struct {
+	ctrlClient client.Client
+	logger     micrologger.Logger
+}
+
+func New(config Config) (*Resource, error) {
+	if config.CtrlClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.CtrlClient must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	r := &Resource{
+		ctrlClient: config.CtrlClient,
+		logger:     config.Logger,
+	}
+
+	return r, nil
+}
+
+// EnsureCreated is a no-op.
+func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	return nil
+}
+
+// EnsureDeleted ensures that all dependent CRs are deleted before finalizer
+// from Cluster is removed.
+func (r *Resource) EnsureDeleted(ctx context.Context, obj interface{}) error {
+	cr, err := key.ToCluster(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	exists, err := r.infrastructureCRExists(ctx, cr)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if exists {
+		finalizerskeptcontext.SetKept(ctx)
+		return nil
+	}
+
+	exists, err = r.machinePoolCRsExist(ctx, cr)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if exists {
+		finalizerskeptcontext.SetKept(ctx)
+		return nil
+	}
```

Comment:
So I could understand better:
If someone tries to delete the `Cluster` resource before deleting the `AzureCluster` and `MachinePool` resources, the deletion will not be finished because we're forcefully keeping the azure-operator finalizer in place. Once the `AzureCluster` and all `MachinePool` resources that belong to the cluster have been deleted, the finalizer will also be removed, so deleting the `Cluster` resource will now finish and the world will be great again.

Is this correct?

Reply:
Yes. Then to dive deeper, similar handler in `MachinePool` controller ensures that `AzureMachinePool` CRs are gone before it can be deleted.